tests/integration/nm/ovs_test.py

```diff
@@ -0,0 +1,185 @@
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from contextlib import contextmanager
+
+import pytest
+
+from libnmstate import nm
+from libnmstate.schema import OVSBridge as OB
+
+from .testlib import mainloop
+
+
+BRIDGE0 = 'brtest0'
+ETH1 = 'eth1'
+
+
+@pytest.fixture
+def bridge_minimum_config():
+    return {
+        OB.CONFIG_SUBTREE: {
+            OB.PORT_SUBTREE: []
+        }
+    }
+
+
+@pytest.fixture
+def bridge_default_config():
+    return {
+        OB.CONFIG_SUBTREE: {
+            OB.OPTIONS_SUBTREE: {
+                OB.FAIL_MODE: '',
+                OB.MCAST_SNOOPING_ENABLED: False,
+                OB.RSTP: False,
+                OB.STP: False
+            },
+            OB.PORT_SUBTREE: []
+        }
+    }
+
+
+def test_create_and_remove_minimum_config_bridge(bridge_minimum_config,
+                                                 bridge_default_config):
+    bridge_desired_state = bridge_minimum_config
+
+    with _bridge_interface(bridge_desired_state):
+        bridge_current_state = _get_bridge_current_state()
+        assert bridge_current_state == bridge_default_config
+        assert bridge_current_state[OB.CONFIG_SUBTREE][OB.PORT_SUBTREE] == []
+
+    assert not _get_bridge_current_state()
+
+
+def test_bridge_with_system_port(eth1_up, bridge_default_config):
+    bridge_desired_state = bridge_default_config
+
+    eth1_port = {
+        OB.PORT_NAME: 'eth1',
+        OB.PORT_TYPE: 'system',
+        # OVS vlan/s are not yet supported.
+        # OB.PORT_VLAN_MODE: None,
+        # OB.PORT_ACCESS_TAG: 0,
+    }
+
+    bridge_desired_state[OB.CONFIG_SUBTREE][OB.PORT_SUBTREE].append(eth1_port)
+
+    with _bridge_interface(bridge_desired_state):
+        bridge_current_state = _get_bridge_current_state()
+        assert bridge_desired_state == bridge_current_state
+
+    assert not _get_bridge_current_state()
+
+
+@contextmanager
+def _bridge_interface(state):
+    try:
+        _create_bridge(state)
+        yield
+    finally:
+        _delete_iface(BRIDGE0)
+        for p in state[OB.CONFIG_SUBTREE][OB.PORT_SUBTREE]:
+            _delete_iface(nm.ovs.PORT_PROFILE_PREFIX + p[OB.PORT_NAME])
+            _delete_iface(p[OB.PORT_NAME])
+
+
+def _get_bridge_current_state():
+    nm.nmclient.client(refresh=True)
+    nmdev = nm.device.get_device_by_name(BRIDGE0)
+    devices_info = [(dev, nm.device.get_device_common_info(dev))
+                    for dev in nm.device.list_devices()]
+    return nm.ovs.get_bridge_info(nmdev, devices_info) if nmdev else {}
```

Comment:
IMHO it is hard to read that the complex lines 103-104 are ignored when nmdev evaluates to False. Would be easiert to read to have a `if nmdev: result = ... else: result = {}; return result`, i.e. the condition first instead at the end.

Reply:
For me, it is easier to read it as is. I have two parameters, and we return one of two options.
If we add the conditions, we will have more code lines, which normally is not a good thing.

Using `x if a else y` is pretty common in python, especially if it lands on a single line.

Reply:
`x if a else y` by itself is not always a problem, however this code is actually:

`nm.ovs.get_bridge_info(nmdev, [(dev, nm.device.get_device_common_info(dev))
                    for dev in nm.device.list_devices()]) if nmdev else {}` because `devices_info` is only used if `nmdev` is considered to be true. Given that `nm.ovs.get_bridge_info()` is just a one-line function to wrap this in a dictionary it seems to be very inconsistent to me to try to put too much complexity in as little lines as possible here. Also I do not agree that more lines of code are a bad thing if the same code is just compressed into less lines.
I guess this code smells again because the absence of a class which seems to require to generate and pass the complex `devices_info` through a lot of methods. Looking more at it, AFAICS `devices_info` does not access any variable in the current scope which makes it even more strange that it actually needs to be generated at this location. `nm.ovs.get_bridge_info()` could just generate the same information by itself.

Nevertheless, I am not going to block the review on this but it does not seemt to be the code quality that you strive for.